lockfile_generator/src/lib.rs

```diff
@@ -0,0 +1,113 @@
+use std::ffi::OsString;
+use std::path::{Path, PathBuf};
+use std::process::{Command, Stdio};
+use std::{fs, io};
+
+pub mod cargo;
+pub mod go;
+pub mod gradle;
+pub mod maven;
+pub mod npm;
+pub mod pip;
+pub mod poetry;
+pub mod python_requirements;
+pub mod yarn;
+
+/// Lockfile generation.
+pub trait Generator {
+    /// Lockfile file name.
+    fn lockfile_name(&self) -> &'static str;
+
+    /// Command for generating the lockfile.
+    fn command(&self) -> Command;
+
+    /// Generate the lockfile for a project.
+    ///
+    /// This will ignore all existing lockfiles and create a new lockfile based
+    /// on the current project configuration.
+    fn generate_lockfile(&self, project_path: &Path) -> Result<String> {
+        // Move existing lockfile so we do not overwrite it.
+        let lockfile_path = project_path.join(self.lockfile_name());
+        let _relocator = if lockfile_path.exists() {
+            Some(FileRelocator::new(lockfile_path.clone())?)
+        } else {
+            None
+        };
+
+        // Generate lockfile at the target location.
+        let mut command = self.command();
+        command.current_dir(project_path);
+        command.stdin(Stdio::null());
+        command.stdout(Stdio::null());
+        command.stderr(Stdio::null());
+
+        // Provide better error message, including the failed program's name.
+        let mut child = command.spawn().map_err(|err| {
+            let program = command.get_program().to_string_lossy().into_owned();
```

Comment:
The `Debug` implementation for `Command` might be more useful here, since it will also print arguments.

From [the docs][1]:

> The default format approximates a shell invocation of the program along with its arguments.

[1]: https://doc.rust-lang.org/std/process/struct.Command.html#impl-Debug-for-Command

```suggestion
            let program = format!("{command:?}");
```